ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java

```diff
@@ -240,53 +242,74 @@ long getMaxBlobSize() {
    * @throws StoreException if BlobStoreStats is not enabled or closed
    */
   Pair<Long, NavigableMap<String, Long>> getValidDataSizeByLogSegment(TimeRange timeRange) throws StoreException {
+    return getValidDataSizeByLogSegment(timeRange, time.milliseconds());
+  }
+
+  /**
+   * Same as {@link #getValidDataSizeByLogSegment(TimeRange)}, but provides a expiry reference time. Blobs with expiration
+   * time less than the given expiry reference time, are considered as expired.
+   * @param timeRange the delete reference {@link TimeRange} at which the data is requested. Defines both the reference time
+   *                  and the acceptable resolution.
+   * @param expiryReferenceTime the reference time for expired blobs. Blobs with expiration time less than it would be
+   *                            considered as expired. Usually it's now.
+   * @return a {@link Pair} whose first element is the time at which stats was collected (in ms) and whose second
+   * element is the valid data size for each segment in the form of a {@link NavigableMap} of segment names to
+   * valid data sizes.
+   * @throws StoreException
+   */
+  Pair<Long, NavigableMap<String, Long>> getValidDataSizeByLogSegment(TimeRange timeRange, long expiryReferenceTime)
+      throws StoreException {
     if (!enabled.get()) {
       throw new StoreException(String.format("BlobStoreStats is not enabled or closing for store %s", storeId),
           StoreErrorCodes.Store_Shutting_Down);
     }
-    ScanResults currentScanResults = scanResults.get();
     Pair<Long, NavigableMap<String, Long>> retValue = null;
+    ScanResults currentScanResults = scanResults.get();
     long referenceTimeInMs = getLogSegmentRefTimeMs(currentScanResults, timeRange);
-    if (referenceTimeInMs != REF_TIME_OUT_OF_BOUNDS) {
-      retValue = currentScanResults.getValidSizePerLogSegment(referenceTimeInMs);
-    } else {
-      if (isScanning && getLogSegmentRefTimeMs(indexScanner.newScanResults, timeRange) != REF_TIME_OUT_OF_BOUNDS) {
-        scanLock.lock();
-        try {
-          if (isScanning) {
-            if (waitCondition.await(waitTimeoutInSecs, TimeUnit.SECONDS)) {
+    if (enableBucketForLogSegmentReports) {
+      if (referenceTimeInMs != REF_TIME_OUT_OF_BOUNDS) {
+        retValue = currentScanResults.getValidSizePerLogSegment(referenceTimeInMs, expiryReferenceTime);
+      } else {
+        if (isScanning && getLogSegmentRefTimeMs(indexScanner.newScanResults, timeRange) != REF_TIME_OUT_OF_BOUNDS) {
+          scanLock.lock();
```

Comment:
I am actually confused about this `scanLock`. If IndexScanner is running (which means `isScanning == true`), current thread is able to acquire and hold `scanLock` (because it can be released by IndexScanner at line 1089). After that, current thread can be blocked by `waitCondition` (without releasing lock), however, IndexScanner has to re-acquire the  lock before signaling the `waitCondition` (see line 1141 - 1146).  It seems like a deadlock except for `waitCondition` can time out eventually. Is this intended?

Reply:
waitCondition is a condition on scanLock, when calling waitCondition.await, it will release the scan lock. This is pretty much how the condition and lock work.

Reply:
Got it, I didn't notice `waitCondition` will release the lock.